src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/Projects/LegacyPackageReferenceProject.cs

```diff
@@ -179,50 +180,63 @@ public override async Task<IEnumerable<PackageReference>> GetInstalledPackagesAs
             return packages.InstalledPackages;
         }
 
-        /// <summary>
-        /// Gets the both the installed (top level) and transitive package references for this project.
-        /// Returns the package reference as two separate lists (installed and transitive).
-        /// </summary>
+        /// <inheritdoc/>
         public override async Task<ProjectPackages> GetInstalledAndTransitivePackagesAsync(CancellationToken token)
         {
-            PackageSpec packageSpec = await GetPackageSpecAsync(NullSettings.Instance);
-
-            var frameworkSorter = new NuGetFrameworkSorter();
-
-            string assetsFilePath = await GetAssetsFilePathAsync();
-            var fileInfo = new FileInfo(assetsFilePath);
-            IList<LockFileTarget> targets = default;
-
-            if (fileInfo.Exists && fileInfo.LastWriteTimeUtc > _lastTimeAssetsModified)
+            PackageSpec packageSpec = await GetCachedPackageSpecAsync(token);
+            if (packageSpec == null)
             {
-                await TaskScheduler.Default;
-                LockFile lockFile = LockFileUtilities.GetLockFile(assetsFilePath, NullLogger.Instance);
-                if (!(lockFile is null))
-                {
-                    targets = lockFile.Targets;
+                IsInstalledAndTransitiveComputationNeeded = true;
 
-                    _lastTimeAssetsModified = fileInfo.LastWriteTimeUtc;
+                return new ProjectPackages(Array.Empty<PackageReference>(), Array.Empty<TransitivePackageReference>());
+            }
 
-                    // clear the transitive packages cache, since we don't know when a dependency has been removed
-                    _transitivePackages.Clear();
-                }
+            IList<LockFileTarget> targetsList = null;
+            if (IsInstalledAndTransitiveComputationNeeded)
+            {
+                // clear the transitive packages cache, since we don't know when a dependency has been removed
+                ClearCache();
+                targetsList = await GetTargetsListAsync(token);
             }
 
+            var frameworkSorter = new NuGetFrameworkSorter();
+
             // get the installed packages
-            IEnumerable<PackageReference> installedPackages = packageSpec
+            List<PackageReference> installedPackages = packageSpec
                .TargetFrameworks
-               .SelectMany(f => GetPackageReferences(f.Dependencies, f.FrameworkName, _installedPackages, targets))
+               .SelectMany(f => GetPackageReferences(
```

Comment:
nit: There's only 1 method difference between legacy and CPS. 

You can probably make that one method abstract and then just keep this logic in the PackageReferenceProject class.

Reply:
Refactored `GetInstalledAndTransitivePackagesAsync()` to get a common implementation in `PackageReference` project class.